code/fmri/load_save.py

```diff
@@ -0,0 +1,367 @@
+import logging
+from typing import Optional, Union
+
+from bids import BIDSLayout
+import numpy as np
+from nibabel import loadsave
+from bids.layout import parse_file_entities
+from bids.layout.writing import build_path
+from nilearn.datasets import fetch_atlas_difumo
+from nilearn.interfaces.fmriprep.load_confounds import _load_single_confounds_file
+
+
+TIMESERIES_PATTERN: list = [
+    "sub-{subject}[/ses-{session}]/func/sub-{subject}"
+    "[_ses-{session}][_task-{task}][_desc-{desc}]"
+    "_{suffix}{extension}"
+]
+TIMESERIES_FILLS: dict = {"desc": "denoised", "extension": ".tsv"}
+CONFOUND_PATTERN: list = [
+    "sub-{subject}[_ses-{session}][_task-{task}][_part-{part}][_desc-{desc}]"
+    "_{suffix}{extension}"
+]
+CONFOUND_FILLS: dict = {"desc": "confounds", "suffix": "timeseries", "extension": "tsv"}
+
+
+def separate_by_similar_values(
+    input_list: list, external_value: Optional[Union[list, np.ndarray]] = None
+) -> dict:
+    """This returns elements of `input_list` with similar values (optionally set by
+    `external_value`) separated into sub-lists.
+
+    Parameters
+    ----------
+    input_list : list
+        List to be separated.
+    external_value : Optional[list], optional
+        Values corresponding to the elements of `input_list`, by default None
+
+    Returns
+    -------
+    dict
+        Dictionnary where each entry is a list of elements that have similar values and
+        the keys are the value for each list.
+    """
+    if external_value is None:
+        external_value = input_list
+
+    data_by_value = defaultdict(list)
+
+    for val, data in zip(external_value, input_list):
+        data_by_value[val].append(data)
+    return data_by_value
+
+
+def get_func_filenames_bids(
+    paths_to_func_dir: str,
+    task_filter: list = [],
+    ses_filter: list = [],
+    run_filter: list = [],
+) -> tuple[list[list[str]], list[float]]:
+    """Return the BIDS functional imaging files matching the specified task and session
+    filters as well as the first (if multiple) unique repetition time (TR).
+
+    Parameters
+    ----------
+    paths_to_func_dir : str
+        Path to the BIDS (usually derivatives) directory
+    task_filter : list, optional
+        List of task name(s) to consider, by default []
+    ses_filter : list, optional
+        List of session name(s) to consider, by default []
+    run_filter : list, optional
+        List of run(s) to consider, by default []
+
+    Returns
+    -------
+    tuple[list[list[str]], list[float]]
+        Returns two lists with: a list of sorted filenames and a list of TRs.
+    """
+    logging.debug("Using BIDS to find functional files...")
+
+    layout = BIDSLayout(
+        paths_to_func_dir,
+        validate=False,
+    )
+
+    all_derivatives = layout.get(
+        scope="all",
+        return_type="file",
+        extension=["nii.gz", "gz"],
+        suffix="bold",
+        task=task_filter,
+        session=ses_filter,
+        run=run_filter,
```

Comment:
Necessary because the defaults are not empty lists anymore. Please note that this implementation disallows setting `None` (which has a meaning to the layout.get)

```suggestion
        task=task_filter or [],
        session=ses_filter or [],
        run=run_filter or [],
```